### ISSUE TITLE
Use Service Discovery

### ISSUE THREAD

**Issue by ignasi35** [akka/akka-grpc#137]
Use Service discovery so `host:port` are not hardcoded.

**Comment by ktoso** [comment]
Relates to managed calls I think... Agree, should be able to plug in some client-side load balancing or similar mechanisms using however we decide to make this pluggable

**Comment by raboof** [comment]
Is this something we'd want to consider when designing the akka-http API's for client-side HTTP2 support? Previously it seemed we'd need little API changes (just implementation), does this mean we should revisit that?

**Comment by ktoso** [comment]
Perhaps, though it's not like HTTP/2 is "new API" really, it should mirror the HTTP one. So it is more about Akka HTTP clientside perhaps needing something. Though in general I think this belongs into the managed calls... Perhaps we need the client report back things to the managed calls, like "errors keep happening on this connection" and the managed calls would decide to try a new host or something.

**Comment by ignasi35** [comment]
Did some investigation recently: https://github.com/ignasi35/akka-grpc-labs

The issue spans into `Channel` pooling.

_NOTE_: I use the name `Channel` since that's the `gRPC` naming convention (beyond language or implementation details) and not because that's the current implementation.

**Comment by ignasi35** [comment]
@raboof I'm not familiar with Akka HTTP client pooling connections but this is highly related. Multiplexing traffic over a `Channel` just adds a bit more complexity since a `Channel` is suddenly used for unary and streamed calls all at once and the lifecycle management gets tricker.

**Comment by ignasi35** [comment]
side note: I poked into this as part of spiking the needs to use akka-grpc in lagom. See more lab work specific to lagom at https://github.com/ignasi35/lagom-grpc-labs

**Comment by ignasi35** [comment]
>  Previously it seemed we'd need little API changes (just implementation), does this mean we should revisit that?

I think so. I think the `ServiceClient` should get a `ServiceDiscovery` instance which could default to [`HardcodedServiceDiscovery`](https://github.com/ignasi35/akka-grpc-labs/blob/master/managed-client/src/main/scala/com/lightbend/akka/labs/utils/HardcodedServiceDiscovery.scala).

Alternatively we could provide more generators:

 * default generator provides a `ServiceClient(Channel)`
 * discovery generator provides a `ServiceClient(...)` which given a `ChannelProvider` (pool, service discovery, circuit breaking,...) [wraps calls](https://github.com/ignasi35/lagom-grpc-labs/blob/1987b5ea45757317a1ac2b734052e1131fcc87f6/hello-impl/src/main/scala/com/example/hello/impl/HelloLoader.scala#L42-L56).

**Comment by ktoso** [comment]
> I think so. I think the ServiceClient should get a ServiceDiscovery instance which could default to HardcodedServiceDiscovery.

Yes, sound right -- to clarify though: that is not Akka HTTP Client API changes though, which was Arnout's question as I understood it. 

The discovery could be provided by the "managed calls" (they really need a working name... this sounds too limiting to what this is aimed to be doing) as it would want to decide based on feedback based on some SPI from the Akka Client that "oh man, that endpoint is failing a lot! Let's try another one". The client here would continue issuing requests via the resolved address -- which would be managed by the managed-calls -- so it would, hopefully, not know anything about the "managing" actually happening. 

Though that's my speculation for now... I hope we can keep them detached as much as possible, but we'll see how it goes
